🚨 Cover middleware/wsgi.py on mypy

Vibhu-Agarwal opened the pull request:
Part of #998 
Extends #1018

I've tried to solve most of the issues, but some still remain - not sure how to solve them.

jkklapp committed:
add types to wsgi module

Vibhu-Agarwal committed:
Merge jkklapp:add_typing_to_wsgi_module

Vibhu-Agarwal committed:
Add middleware/wsgi.py to setup.cfg

Vibhu-Agarwal committed:
Smoothen out a few (mypy-related) bugs

I don't know how to solve the remaining ones

Vibhu-Agarwal committed:
Small Fix: use HTTPRequestEvent

Vibhu-Agarwal commented:
Not sure if this commit ^^^ is correct.
Will revert, if it's not.

Vibhu-Agarwal commented on uvicorn/middleware/wsgi.py:
```diff
@@ -99,11 +100,11 @@ def __init__(
     ) -> None:
-        message = await receive()
+        message: HTTPRequestEvent = await receive()
         body = message.get("body", b"")
```
Before the change, it was throwing this error:
```
uvicorn/middleware/wsgi.py:103: error: TypedDict "HTTPDisconnectEvent" has no key 'body'  [misc]
uvicorn/middleware/wsgi.py:103: error: TypedDict "WebsocketConnectEvent" has no key 'body'  [misc]
uvicorn/middleware/wsgi.py:103: error: TypedDict "WebsocketReceiveEvent" has no key 'body'  [misc]
uvicorn/middleware/wsgi.py:103: error: TypedDict "WebsocketDisconnectEvent" has no key 'body'  [misc]
uvicorn/middleware/wsgi.py:103: error: TypedDict "LifespanStartupEvent" has no key 'body'  [misc]
uvicorn/middleware/wsgi.py:103: error: TypedDict "LifespanShutdownEvent" has no key 'body'  [misc]
```

Kludex replied:
The `http.disconnect` never gets here?

Vibhu-Agarwal replied:
That's exactly what was bugging me 😅
Most probably `http.disconnect` could get here.
But if I remove the type annotation from `message`, the `body = message.get("body", b"")` line throws out the errors.
Not sure how to resolve that then.


PS. The same goes for `body_message` a couple of lines below that.

Kludex commented on uvicorn/middleware/wsgi.py:
```diff
@@ -46,43 +59,52 @@ def build_environ(scope, message, body):
 
-    async def __call__(self, scope, receive, send):
+    async def __call__(
+        self, scope: HTTPScope, receive: ASGIReceiveCallable, send: ASGISendCallable
```
What's the error you have below? 
```suggestion
        self, scope: Scope, receive: ASGIReceiveCallable, send: ASGISendCallable
```

Vibhu-Agarwal replied:
There isn't any error in this line.
Also, I think `HTTPScope` should be fine as it mentions this one line down:
https://github.com/encode/uvicorn/blob/a8af0da26c921b27849701f2cedbf89b8f3cbd86/uvicorn/middleware/wsgi.py#L62

Kludex replied:
Yeah, but the line below says that it should be `HTTPScope` **after** it, not before. Otherwise, we'd not need to assert it.

Vibhu-Agarwal replied:
The `assert` has been put just after the function header - in the first line.
https://github.com/encode/uvicorn/blob/a8af0da26c921b27849701f2cedbf89b8f3cbd86/uvicorn/middleware/wsgi.py#L61-L62
My reasoning was that this should mean that the function is expecting the type of `scope` as HTTP, and therefore I put the typehint for that.

Also, I tried to replace that with `Scope`. That threw new mypy errors as the rest of the code is expecting attributes on `scope` which are only there in `HTTPScope` and not in `Scope`.

Kludex commented on uvicorn/middleware/wsgi.py:
```diff
@@ -46,43 +59,52 @@ def build_environ(scope, message, body):
+    def __init__(
+        self,
+        app: ASGI3Application,
+        executor: concurrent.futures.ThreadPoolExecutor,
```
```suggestion
        executor: concurrent.futures.Executor,
```

Kludex commented on uvicorn/middleware/wsgi.py:
```diff
@@ -36,8 +49,8 @@ def build_environ(scope, message, body):
     # Go through headers and make them into environ entries
-    for name, value in scope.get("headers", []):
-        name = name.decode("latin1")
+    for name_, value_ in scope.get("headers", []):
```
Can we use `key, value` instead? And on the `value` below, `value_str`? Or just maintain the same `name, value` here, but using the `name_str` and `value_str` below?

Kludex reviewed (commented):
Thanks for the PR @Vibhu-Agarwal ! :smiley: 

I've added some comments, can you check them, please?

Vibhu-Agarwal committed:
Apply suggestions from PR review

euri10 commented:
I've not looked at this but you can maybe take inspiration from https://github.com/encode/uvicorn/pull/1049 where I typed everything in the same module I think

Vibhu-Agarwal committed:
Remove all mypy issues

Taking ideas from: https://github.com/encode/uvicorn/blob/dd85cdacf154529ea1c3d12b5bda7808673979f2/uvicorn/middleware/wsgi.py

euri10 commented on uvicorn/middleware/wsgi.py:
```diff
@@ -111,31 +135,43 @@ async def sender(self, send):
-            self.send_queue.append(
-                {"type": "http.response.body", "body": chunk, "more_body": True}
-            )
+    def wsgi(self, environ: Dict, start_response: Awaitable[None]) -> None:
```
Once https://github.com/encode/uvicorn/pull/1067 is merged (which is soon !) there will be a _types.py you can use for typing this

Vibhu-Agarwal committed:
Merge branch 'encode:master' into jkl_1018

Vibhu-Agarwal committed:
Adjusting w.r.t. to #1067

Vibhu-Agarwal committed:
Adjusting w.r.t. to #1067

Kludex commented on uvicorn/middleware/wsgi.py:
```diff
@@ -37,52 +52,63 @@ def build_environ(scope, message, body):
         if corrected_name in environ:
-            value = environ[corrected_name] + "," + value
-        environ[corrected_name] = value
+            corrected_name_environ = environ[corrected_name]
```
This looks a little too much just to satisfy `mypy` :thinking:

Kludex replied:
Should we just ignore it here?

euri10 replied:
lgtm

Kludex commented on uvicorn/middleware/wsgi.py:
```diff
@@ -111,31 +137,43 @@ async def sender(self, send):
-            status_code, _ = status.split(" ", 1)
-            status_code = int(status_code)
+            status_code_str, _ = status.split(" ", 1)
+            status_code: int = int(status_code_str)
```
Why do we need to add the annotation here? :thinking:

Kludex replied:
Ah, I guess we needed it before the renaming.

Kludex reviewed (commented):
We're almost there! I've added some comments, do you mind taking a look @Vibhu-Agarwal ?

Thanks for your help so far!

euri10 commented on setup.cfg:
```diff
@@ -23,6 +23,7 @@ files =
     uvicorn/protocols/websockets/auto.py,
     uvicorn/supervisors/__init__.py,
     uvicorn/middleware/debug.py,
+    uvicorn/middleware/wsgi.py,
```
```suggestion
    uvicorn/middleware/wsgi.py,
    tests/middleware/test_wsgi.py,
```
let's add the test also, this will show you some easy to spot errors in the types

euri10 commented on uvicorn/middleware/wsgi.py:
```diff
@@ -37,52 +52,63 @@ def build_environ(scope, message, body):
 
 class WSGIMiddleware:
-    def __init__(self, app, workers=10):
+    def __init__(self, app: ASGI3Application, workers: int = 10):
```
this is wrong, app here is not an asgi application but a wsgi application:
`WSGIApp = typing.Callable[[Environ, StartResponse], typing.Iterable[bytes]]`

euri10 replied:
can be added to _types.py

euri10 commented on uvicorn/middleware/wsgi.py:
```diff
@@ -37,52 +52,63 @@ def build_environ(scope, message, body):
-    def __init__(self, app, executor, scope):
+    def __init__(
+        self,
+        app: ASGI3Application,
```
same here, WSGIApp

euri10 commented on uvicorn/middleware/wsgi.py:
```diff
@@ -37,52 +52,63 @@ def build_environ(scope, message, body):
+    def __init__(
+        self,
+        app: ASGI3Application,
+        executor: concurrent.futures.Executor,
```
it is a `ThreadPoolExecutor`

euri10 commented on uvicorn/middleware/wsgi.py:
```diff
@@ -37,52 +52,63 @@ def build_environ(scope, message, body):
-        self.send_queue = []
-        self.loop = None
+        self.send_queue: List[Optional[ASGISendEvent]] = []
+        self.loop: asyncio.AbstractEventLoop = None  # type: ignore
```
no need to ignore that, we can instantiate it directly

euri10 commented on uvicorn/middleware/wsgi.py:
```diff
@@ -37,52 +52,63 @@ def build_environ(scope, message, body):
+    async def __call__(
+        self, receive: ASGIReceiveCallable, send: ASGISendCallable
+    ) -> None:
+        message: HTTPRequestEvent = await receive()  # type: ignore
```
worth maybe putting a note as to why we ignore this, iirc because mypy cannot infer the Literal well enough, or put the [assignment] so that we know

Vibhu-Agarwal committed:
Applied suggestions from PR review

euri10 reviewed (approved):
this lgtm @Vibhu-Agarwal thanks a lot
I let @Kludex finish it :)

edit: oups didnt see the CI was failing, will have to wait >_)

euri10 commented:
you can use this patch @Vibhu-Agarwal I think this solves the rest

```patch
Index: tests/middleware/test_wsgi.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/tests/middleware/test_wsgi.py b/tests/middleware/test_wsgi.py
--- a/tests/middleware/test_wsgi.py	(revision 7610a04a8be1219b2b48eff20e73ad0049884c45)
+++ b/tests/middleware/test_wsgi.py	(revision 6b84705a15ace9d662a49f15921c54bdfd8287c1)
@@ -3,7 +3,7 @@
 
 import httpx
 import pytest
-from asgiref.typing import HTTPScope
+from asgiref.typing import HTTPRequestEvent, HTTPScope
 
 from uvicorn._types import Environ, StartResponse
 from uvicorn.middleware.wsgi import WSGIMiddleware, build_environ
@@ -16,7 +16,7 @@
         ("Content-Type", "text/plain; charset=utf-8"),
         ("Content-Length", str(len(output))),
     ]
-    start_response(status, headers)
+    start_response(status, headers, None)
     return [output]
 
 
@@ -27,7 +27,7 @@
         ("Content-Type", "text/plain; charset=utf-8"),
         ("Content-Length", str(len(output))),
     ]
-    start_response(status, headers)
+    start_response(status, headers, None)
     return [output]
 
 
@@ -45,7 +45,7 @@
             ("Content-Type", "text/plain; charset=utf-8"),
             ("Content-Length", str(len(output))),
         ]
-        start_response(status, headers, exc_info=sys.exc_info())
+        start_response(status, headers, sys.exc_info())  # type: ignore[arg-type]
         return [output]
 
 
@@ -108,7 +108,12 @@
         "root_path": "/文",
         "query_string": b"a=123&b=456",
         "headers": [(b"key", b"value1"), (b"key", b"value2")],
+    }  # type: ignore[typeddict-item]
+    message: HTTPRequestEvent = {
+        "type": "http.request",
+        "body": b"",
+        "more_body": False,
     }
-    environ = build_environ(scope, b"", b"")
+    environ = build_environ(scope, message, b"")
     assert environ["PATH_INFO"] == "/文".encode("utf8").decode("latin-1")
     assert environ["HTTP_KEY"] == "value1,value2"
Index: uvicorn/_types.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/uvicorn/_types.py b/uvicorn/_types.py
--- a/uvicorn/_types.py	(revision 7610a04a8be1219b2b48eff20e73ad0049884c45)
+++ b/uvicorn/_types.py	(revision 6b84705a15ace9d662a49f15921c54bdfd8287c1)
@@ -9,4 +9,6 @@
 StartResponse = typing.Callable[
     [str, typing.Iterable[typing.Tuple[str, str]], typing.Optional[ExcInfo]], None
 ]
-WSGIApp = typing.Callable[[Environ, StartResponse], typing.Iterable[bytes]]
+WSGIApp = typing.Callable[
+    [Environ, StartResponse], typing.Union[typing.Iterable[bytes], BaseException]
+]

```

Vibhu-Agarwal committed:
Fixed remaining test_wsgi.py mypy issues

Vibhu-Agarwal commented:
Failing on **3.6** and **3.7** 🤔

euri10 commented:
typeddict is > 3.7 just need to fix the import

On Wed, Jun 23, 2021 at 9:22 AM Vibhu Agarwal ***@***.***>
wrote:

> Failing on 3.7 and 3.8 🤔
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/encode/uvicorn/pull/1075#issuecomment-866596061>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAINSPSUE25MMWUZBNDTE2LTUGDRXANCNFSM46MVJV5A>
> .
>

Vibhu-Agarwal committed:
Trying by removing "[typeddict-item]" beside "type: ignore"

Vibhu-Agarwal committed:
Revert previous commit

Vibhu-Agarwal commented:
Unable to figure out the fix 😕
Import doesn't seem like the issue here, as it's coming directly from `asgiref` where they're checking the version and aptly importing from `typing`/`typing_extensions`.
For some reason, `# type: ignore` isn't working 🤔

euri10 commented:
you can fill the scope too
```
    scope: HTTPScope = {
        "type": "http",
        "asgi": {"version": "3.0", "spec_version": "2.0"},
        "scheme": "http",
        "http_version": "1.1",
        "method": "GET",
        "path": "/文",
        "raw_path": b"/\xe6\x96\x87",
        "root_path": "/文",
        "query_string": b"a=123&b=456",
        "client": None,
        "server": None,
        "headers": [(b"key", b"value1"), (b"key", b"value2")],
        "extensions": {},
    }
```

Vibhu-Agarwal committed:
Final fix ... yaayyy!

Vibhu-Agarwal commented:
Thanks for the reviews and support 👍
Cc @euri10 @Kludex

euri10 commented:
thanks a lot @Vibhu-Agarwal !
